services/src/main/scala/cromwell/services/metadata/impl/builder/MetadataBuilderActor.scala

```diff
@@ -488,6 +515,134 @@ class MetadataBuilderActor(readMetadataWorkerMaker: () => Props,
+      )
+    }
+  }
+
+  private def extractFromJsAs[A: Manifest](js: JsObject, fieldName: String): Option[A] =
+    js.fields.get(fieldName) match {
+      case Some(a: A) => Some(a)
+      case _ => None
+    }
+
+  private def computeCost(callName: String, jsVal: JsValue): ErrorOr[BigDecimal] =
+    jsVal match {
+      case jsCall: JsObject =>
+        extractFromJsAs[JsString](jsCall, CallMetadataKeys.VmStartTime) map { startTimeVal =>
+          val startTimeErrorOr = ErrorOr(OffsetDateTime.parse(startTimeVal.value))
+
+          val endTimeErrorOr =
+            extractFromJsAs[JsString](jsCall, CallMetadataKeys.VmEndTime)
+              .map(v => ErrorOr(OffsetDateTime.parse(v.value)))
+              .getOrElse(OffsetDateTime.now().validNel)
+
+          val rawCostPerHour = extractFromJsAs[JsNumber](jsCall, CallMetadataKeys.VmCostPerHour).map(_.value)
+          val costPerHourErrorOr = rawCostPerHour match {
+            case Some(c: BigDecimal) if c >= 0 => c.validNel
+            // A costPerHour < 0 indicates an error
+            case Some(_: BigDecimal) =>
+              val index =
+                extractFromJsAs[JsNumber](jsCall, "shardIndex").map(_.value.toString).getOrElse("-1")
+              val attempt =
+                extractFromJsAs[JsNumber](jsCall, "attempt").map(_.value.toString).getOrElse("1")
+              s"Couldn't find valid vmCostPerHour for ${List(callName, index, attempt).mkString(".")}".invalidNel
+            case None => BigDecimal(0).validNel
+          }
+
+          for {
+            start <- startTimeErrorOr
+            end <- endTimeErrorOr
+            costPerHour <- costPerHourErrorOr
+            vmRuntimeInMillis = start.until(end, ChronoUnit.MILLIS).toDouble
+          } yield (vmRuntimeInMillis / (1000 * 60 * 60)) * costPerHour
+        } getOrElse (BigDecimal(0).validNel)
+      // TODO is millis the right thing to use here? Should it be seconds?
+      case _ => BigDecimal(0).validNel
+    }
+
+  def buildCostAndStop(id: WorkflowId,
+                       status: String,
+                       eventsList: Seq[MetadataEvent],
+                       expandedValues: Map[String, JsValue],
+                       target: ActorRef,
+                       originalRequest: BuildMetadataJsonAction
+  ): State = {
+
+    val metadataEvents = MetadataBuilderActor.parse(groupEvents(eventsList), expandedValues)
+
+    // Walk the structured metadata to attempt to compute cost for each call in this workflow
+    val callCostsWithErrors: List[ErrorOr[BigDecimal]] = for {
+      wfObj <- extractFromJsAs[JsObject](metadataEvents, id.toString).toList
+      callsObj <- extractFromJsAs[JsObject](wfObj, "calls").toList
+      singleCallName <- callsObj.fields.keys.toList
+      singleCallAttemptList <- extractFromJsAs[JsArray](callsObj, singleCallName).toList
+      singleCallAttempt <- singleCallAttemptList.elements.toList
+      singleCallAttemptCost = computeCost(singleCallName, singleCallAttempt)
```

Comment:
What is your current thinking about how we should handle calls with multiple attempts?

Reply:
I think they should be handled correctly here, it should sum the costs for all of the attempts.